### PR TITLE
Kwxm/plc builtin updates

### PR DESCRIPTION
**Pull request by kwxm**
This updates the syntax for builtin application to avoid ambiguity between type and term variables, hopefully dealing with #1143.  The syntax is implemented in the branch for saturated builtins.

Also, it adds a missing hypothesis about kinds to the typing rule for builtin applications.

### CODE REVIEW

File: plutus-core-spec/figures/TypeSynthesis.tex [hunk 1]
```diff
@@ -202,12 +202,13 @@
```
Comment (kwxm) [inline plutus-core-spec/figures/TypeSynthesis.tex:202]: I don't think this file is used.  The typing rules are included directly in plutus-core-specification.tex now. I'm nit sure how that happened.

File: plutus-core-spec/figures/TypeSynthesis.tex [hunk 2]
```diff
@@ -202,12 +202,13 @@
 
     \begin{prooftree}
         \alwaysNoLine
-        \AxiomC{$bn$ has signature $\sig{\alpha_0 :: K_0, ..., \alpha_m :: K_m}{B_0, ..., B_n}{C}$ in Figure \ref{fig:builtins}}
+        \AxiomC{$bn$ has signature $\sig{\alpha_1 :: K_1, ..., \alpha_k :: K_k}{B_1, ..., B_n}{C}$ in Figure \ref{fig:builtins}}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istypeJ{A_i}{K_i}}\)}
```
Comment (kwxm) [inline plutus-core-spec/figures/TypeSynthesis.tex:206]: Do we want everything to be indexed by `i` here?  It kind of makes it looks as if things have to match up that actually don't have to.

**Comment by kwxm** [comment]
While I'm here, I notice that in the typed machines we create a frame for type abstractions and then immediately discard it.  Should we just have 
```
s ▻ {M A} ↦ s ▻ M
```
instead?  I have a vague memory that we used to do something different here and then changed it.

**Comment by effectfully** [comment]
> Should we just have `s ▻ {M A} ↦ s ▻ M` instead?

There's a corner case: when `M` is a builtin and it doesn't get fully saturated in the end, we return the full `{M A}` application as a result. Not sure if it's important, though, as we do not keep track of type variables and stuff in the CEK machine, so that `A` can be ill-scoped anyway.

I'd perhaps wait until we have saturated builtins and then change the evaluation semantics the way you suggest.

**Comment by kwxm** [comment]
> There's a corner case: when `M` is a builtin and it doesn't get fully saturated in the end, we return the full `{M A}` application as a result. Not sure if it's important, though, as we do not keep track of type variables and stuff in the CEK machine, so that `A` can be ill-scoped anyway.
False alarm. I realise now that I was talking nonsense: we _do_ need the `{}` frame, because otherwise there's no way to know when an abstraction has to be instantiated, and that matters now that we don't have the value restriction.

**Comment by effectfully** [comment]
> False alarm. I realise now that I was talking nonsense: we do need the `{}` frame, because otherwise there's no way to know when an abstraction has to be instantiated, and that matters now that we don't have the value restriction.

Ah, indeed, we need to drop the `TyAbs`, but for that we need to evaluate `M` first, which means that we need to put the `{}` frame onto the stack. And since `TyAbs` blocks evaluation and we don't have value restriction, we can't just drop it arbitrarily.

File: plutus-core-spec/figures/TypeSynthesis.tex (hunk 2 above)
Comment (michaelpj) [reply plutus-core-spec/figures/TypeSynthesis.tex:206]: I think that's important, no? It's showing that there's a typing hypothesis for all each individual argument.

Comment (kwxm) [reply plutus-core-spec/figures/TypeSynthesis.tex:206]: Yeah, I'm just not quite sure what the convention is, like whether you regard the things on each line as being independent from each other, and whether all of the indices are implicitly universally quantified.  Paging @jmchapman !

**Comment by kwxm** [comment]
The definition of the CEK machine in the document is wrong.  If you give it a term like `[(lam n (con integer) (builtin addInteger n n)) (con 5)]` then it fails because the environment is empty by the time it gets to the second `n` when it's trying to apply `addInteger`.  I think the solution to this is to save the current environment in the frame for `builtin` and then use that to evaluate all of the arguments.  

Since this branch is still open I'll fix that here once I've convinced myself the fix is correct.

**Comment by kwxm** [comment]
I think this should be correct now, but it would be helpful if someone could check.  The problem was that when you evaluate a variable the current environment is discarded and replaced with the one associated with the variable.  When evaluating a builtin it now saves the environment in effect when the evaluation starts and uses it to evaluate all of the arguments (and the builtin itself when the arguments have all been computed, but that's irrelevant at the moment because builtins don't have access to environments).

**Comment by kwxm** [comment]
Nobody's complaining, so I'm just going to merge this. I'd still appreciate confirmation that the update to the way the CEK deals with environments during builtin application looks OK though.

**Comment by jmchapman** [comment]
I plan to start implementing the CEK machine following the spec next week.

**Pull request merged**